The following scripts takes files generated by the main `MetaGPA` pipeline for extensive analysis.

---

## Phylogenetics

`phylogenetics.py` takes the following files from the main `MetaGPA` pipeline:

- `all_contigs_pfam_a.gtf`
- `modified_contigs_pfam_a.gtf`
- `unmodified_contigs_pfam_a.gtf`
- `modified_contigs.fa`
- `unmodified_contigs.fa`

### Requirements

- muscle
- RAxML

### Usage

```bash
python phylogenetics.py --pfam <pfam_name>
```

where the `pfam_name` can be any pfam name of interest

### Output Files

1. `<pfam_name>.faa` pfam fasta file
2. `<pfam_name>.index` index of contigs
3. `RAxML_*.<pfam_name>.tree` raxml output files

---

## Co-occurrence

`co_occurrence.py` takes the following files from the main `MetaGPA` pipeline:

- `modified_contigs_pfam_a.gtf`
- `pfam_a_statistics.csv`

3. package installed on R:
cooccur
   
### Usage

```bash
python co_occurrence.py --number 20
R < co_occurrence.R --no-save 
```

Other options:

```bash
--number  generate a matrix file for the top chosen number of enriched pfams as listed in the pfam_a_statistics.csv file, int, default 20
```

### Output Files
1. `enriched_pfam_a_matrix.csv` file to be used in co_occurrence.R 
2. `enriched_pfam_prob_table.csv` cooccur output file      
3. `significant_negative_correlation.csv` positive correlations with p < 0.05
4. `significant_positive_correlation.csv` negative correlations with p < 0.05
